### ISSUE TITLE
onion socket 9735 is closed when restarting Tor

### ISSUE THREAD

**Issue by holgern** [raspiblitz/raspiblitz#1604]
I had the problem that the tor socket 9735 was not reachable. I checked it by installing
```
sudo apt-get install proxychains4 nmap
```
on a server and checked the port with:
```
proxychains4 nmap -Pn -p 9735 -sT -v my_onion_addr.onion
```
The command returned a 
```
PORT     STATE SERVICE
9735/tcp closed unknown
```

Only a 
```
./config.scripts/internet.tor.sh off
```
reboot
```
./config.scripts/internet.tor.sh on
```
reboot
did repair the port.
Now I see:
```
PORT     STATE SERVICE
9735/tcp open  unknown
```

It is also possible to do the proxychains4 command on the raspi itself. 

Restarting the tor service and removing the ethernet cable did not helped.

**Comment by holgern** [comment]
After testing, I found out that the port 9735 is closed every time tor is restarted due to installing something on the raspiblitz.

**Comment by holgern** [comment]
I found the fix:
lnd must be stopped before tor is restarted

```
sudo systemctl stop lnd
sudo systemctl restart tor@default
sudo systemctl start lnd
```

**Comment by openoms** [comment]
you can get the LND HS address with:
`echo "$(lncli getinfo)" | grep "uris" -A 1 | tr -d '\n' | cut -d '"' -f4 | cut -d @ -f2 | cut -d : -f1`
so testing with:
`proxychains4 nmap -Pn -p 9735 -sT -v $(echo "$(lncli getinfo)" | grep "uris" -A 1 | tr -d '\n' | cut -d '"' -f4 | cut -d @ -f2 | cut -d : -f1)`

It always gives me:
```
PORT     STATE  SERVICE
9735/tcp closed unknown
```

I often find that if some of my peers are disconnecting restarting tor (and often lnd too) helps.
It is sufficient to do:
```
sudo systemctl restart tor
sudo systemctl restart lnd
```

**Comment by holgern** [comment]
I'm getting a denied when I only restart tor.
```
:9735 <--denied
```
This is the problem, when installing something on the raspiblitz. 

I don't know if this happens only to my installation, or if this is a  general problem.
I'm on the current patched github for branch 1.6 (release 1.6.1), so I have included the PR #1571 (I did all the necessary changes on my system)
A 
```
sudo systemctl restart tor@default
sudo systemctl restart lnd
```
fixes the closed port.

**Comment by openoms** [comment]
Thank you for documenting this, it is an interesting find, don't know what to make of it yet, but certainly shows what could have been  the problem with a problematic Tor node of mine which has been hardly restarted ever.
While testing now I could only get 
```
PORT     STATE SERVICE
9735/tcp open  unknown
```
after restarting the whole node and waiting for LND to reconnect the peers.
@frennkie @rootzoll there could be a new way test for the occasional IP2Tor problems as well. 
The internal port of the Hidden Service might be closed in some occasions when restarting Tor?

**Comment by rootzoll** [comment]
Hmm interesting .. let me know if there is anything to add to the v1.6.1 on short notice. Otherwise I mark this for v1.7 to take a deeper look.

**Comment by rootzoll** [comment]
@openoms is this still relevant?

**Comment by openoms** [comment]
don't think so, a lot has been changed. Could reopenif comes up again.

**Comment by rootzoll** [comment]
So closing for now - please reopen is something shows up.

**Comment by openoms** [comment]
I can reproduce this still, suggest to introduce a manual check for it in the CONNECT menu https://github.com/rootzoll/raspiblitz/issues/2090

This causes long channel downtimes and problem using Pool so will take up with LND.

**Comment by openoms** [comment]
Added monitoring for this problem in #2137.

As for now it seems there is no other way to mitigate this only to restart LND with Tor as well. 
This would add unnecessary downtime and more handling of the passwordC.

One partila solution could be to avoid manula restarts at least is to set up a separate Tor instance for LND, so it is not restarted by the `internet.hiddenservice.sh` when new .onion services are added or removed.

How to create muliple Tor instances: https://coderwall.com/p/01_gpg/multiple-instance-of-tor
LND Tor docs: https://github.com/lightningnetwork/lnd/blob/master/docs/configuring_tor.md

**Comment by openoms** [comment]
Can test the dedicated tor instance (in #2148) with these:
Learned from:  https://www.torservers.net/wiki/setup/server#multiple_tor_processes

```
    NODENAME="lnd"
    SOCKSPORT=9070
    CONTROLPORT=$((SOCKSPORT+1))
    echo "# Creating a dedicated Tor instance for $NODENAME"
    sudo tor-instance-create $NODENAME

    echo "# Make sure the user bitcoin is in the _tor-$NODENAME group"
    sudo usermod -a -G _tor-$NODENAME bitcoin

    # create tor data directory if it not exist
    if [ ! -d "/mnt/hdd/tor-$NODENAME" ]; then
      echo "# - creating tor data directory"
      sudo mkdir -p /mnt/hdd/tor-$NODENAME
      sudo mkdir -p /mnt/hdd/tor-$NODENAME/sys
    else
      echo "# - /mnt/hdd/tor-$NODENAME data directory exists"
    fi
    # make sure its the correct owner
    sudo chmod -R 700 /mnt/hdd/tor-$NODENAME
    sudo chown -R _tor-$NODENAME:_tor-$NODENAME /mnt/hdd/tor-$NODENAME

    echo "
### torrc for tor@$NODENAME
### https://github.com/lightningnetwork/lnd/blob/master/docs/configuring_tor.md

DataDirectory /mnt/hdd/tor-$NODENAME/sys
PidFile /mnt/hdd/tor-$NODENAME/sys/tor.pid

SocksPort $SOCKSPORT
ControlPort $CONTROLPORT
CookieAuthentication 1
CookieAuthFileGroupReadable 1

SafeLogging 1
Log notice stdout
Log notice file /mnt/hdd/tor-$NODENAME/notice.log
Log info file /mnt/hdd/tor-$NODENAME/info.log
" | sudo tee /etc/tor/instances/$NODENAME/torrc
    sudo chmod 644 /etc/tor/instances/$NODENAME/torrc

    sudo mkdir -p /etc/systemd/system/tor@$NODENAME.service.d
    sudo tee /etc/systemd/system/tor@$NODENAME.service.d/raspiblitz.conf >/dev/null <<EOF
    # DO NOT EDIT! This file is generated by raspiblitz and will be overwritten
[Service]
ReadWriteDirectories=-/mnt/hdd/tor-$NODENAME
[Unit]
After=network.target nss-lookup.target mnt-hdd.mount
EOF

    echo "Setup logrotate"
    # add logrotate config for modified Tor dir on ext. disk
    sudo tee /etc/logrotate.d/raspiblitz-tor-$NODENAME >/dev/null <<EOF
/mnt/hdd/tor-$NODENAME/*log {
        daily
        rotate 5
        compress
        delaycompress
        missingok
        notifempty
        create 0640 _tor-$NODENAME _tor-$NODENAME
        sharedscripts
        postrotate
                if invoke-rc.d tor status > /dev/null; then
                        invoke-rc.d tor reload > /dev/null
                fi
        endscript
}
EOF
    sudo systemctl daemon-reload
    sudo systemctl enable tor@$NODENAME
    sudo systemctl start tor@$NODENAME
    
    # config for lnd self-test
    echo "
strict_chain
proxy_dns 
tcp_read_time_out 15000
tcp_connect_time_out 8000
[ProxyList]
socks5 	127.0.0.1 $SOCKSPORT
" | sudo tee /etc/proxychains.$NODENAME.conf

    # modify LND service
    echo "# Make sure LND is disabled"
    sudo systemctl disable lnd 2>/dev/null

    echo "# Editing /etc/systemd/system/lnd.service"
    sudo sed -i "s/^ExecStart=\/usr\/local\/bin\/lnd.*\
/ExecStart=\/usr\/local\/bin\/lnd --tor\.active --tor\.streamisolation --tor\.v3 --tor\.socks=$SOCKSPORT --tor\.control=$CONTROLPORT --listen=127\.0\.0\.1\:9735 \${lndExtraParameter}/g" \
    /etc/systemd/system/lnd.service

    echo "# Enable LND again"
    sudo systemctl enable lnd
    echo "# OK"
    echo 
```    
```
# restart lnd
sudo systemctl restart lnd
# unlock
lncli unlock
# self test
proxychains4 nmap -p9735 $(lncli getinfo|jq -c .uris|cut -d@ -f2|cut -d: -f1)

**Comment by rootzoll** [comment]
OK PR #2148 merged - will be part of v1.7RC2 ... up for final testing.

**Comment by nyxnor** [comment]
Ok, now there is 2 instances.
The problem was happening because Tor was restarting, but guess what, it doesnt need to restart when editing the torrc.
https://github.com/rootzoll/raspiblitz/issues/2409#issuecomment-876340098
Adding another instance gives some complication handling bridges that needs a workaround, deleting and restoring lines.
Can this be tested on a v1.6.3 Raspiblitz with reloading tor instead of restarting? Just to see if this issue could be fixed without another instance.
I think adding one reload line keep it more simple than setting a new tor instance although different tor instances could have it place in the future.

@holgern Are you still on v.1.6.3?
@openoms your comment is also important. (also, why use another user _tor-lnd

**Comment by openoms** [comment]
`_tor-lnd` is set up automatically by tor when the new instance is created same way as `debian-tor` is the default.

The main question is: 
can we activate a Hidden Service without needing to restart and unlock LND?

Can test on v1.7.0 as well. Try reloading `tor@lnd` (not sure what you referring to by reloading yet) instead of restarting (aka
 `sudo systemctl restart tor@lnd`) and test if the LND Hidden Service on 9735 is still connectable.

One step test:
```
sudo apt install proxychains4 nmap
proxychains4 nmap -p9735 $(lncli getinfo|jq -c .uris|cut -d@ -f2|cut -d: -f1)
```

**Comment by nyxnor** [comment]
This is the state now, I tried restarting tor@lnd, tor@default, lnd and nothing changed. Also, it is referring to port 9050, it should goes for the 9070 for tor@lnd.
```
proxychains4 nmap -p9735 $(lncli getinfo|jq -c .uris|cut -d@ -f2|cut -d: -f1)
```
>[proxychains] config file found: /etc/proxychains4.conf
[proxychains] preloading /usr/lib/aarch64-linux-gnu/libproxychains.so.4
[proxychains] DLL init: proxychains-ng 4.13
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  []:80 <--denied
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  []:9735 <--denied

>PORT     STATE  SERVICE
9735/tcp closed unknown


lnd wallet is unlocked, lnd is active, now testing editing torrc and reloading tor.
So check if it is unlocked, reload tor and check the lnd logs again.
Did not see LND fail, could you please try this @openoms ?
```
sudo tail -n 10 /mnt/hdd/lnd/logs/bitcoin/mainnet/lnd.log
sudo pkill -sighup tor
sudo tail -n 10 /mnt/hdd/lnd/logs/bitcoin/mainnet/lnd.log
```
You can see that tor@lnd was reloaded with
```
sudo systemctl status tor@lnd
sudo journalctl -eu tor@lnd
```

**Comment by openoms** [comment]
OK so reloading with `sudo pkill -sighup tor` did not break the LND port as `sudo systemctl restart tor@lnd` did so this is awesome: 

```
admin@bob:~ $ proxychains4 nmap -p9735 $(lncli getinfo|jq -c .uris|cut -d@ -f2|cut -d: -f1)
[proxychains] config file found: /etc/proxychains4.conf
[proxychains] preloading /usr/lib/aarch64-linux-gnu/libproxychains.so.4
[proxychains] DLL init: proxychains-ng 4.13
Starting Nmap 7.70 ( https://nmap.org ) at 2021-07-13 14:10 BST
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion:80 <--denied
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion:9735  ...  OK
Nmap scan report for a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion (224.0.0.1)
Host is up (2.5s latency).
rDNS record for 224.0.0.1: all-systems.mcast.net

PORT     STATE SERVICE
9735/tcp open  unknown

Nmap done: 1 IP address (1 host up) scanned in 3.64 seconds

admin@bob:~ $ sudo pkill -sighup tor

admin@bob:~ $ proxychains4 nmap -p9735 $(lncli getinfo|jq -c .uris|cut -d@ -f2|cut -d: -f1)
[proxychains] config file found: /etc/proxychains4.conf
[proxychains] preloading /usr/lib/aarch64-linux-gnu/libproxychains.so.4
[proxychains] DLL init: proxychains-ng 4.13
Starting Nmap 7.70 ( https://nmap.org ) at 2021-07-13 14:11 BST
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion:80 <--denied
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion:9735  ...  OK
Nmap scan report for a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion (224.0.0.1)
Host is up (1.2s latency).
rDNS record for 224.0.0.1: all-systems.mcast.net

PORT     STATE SERVICE
9735/tcp open  unknown

Nmap done: 1 IP address (1 host up) scanned in 1.79 seconds

admin@bob:~ $ sudo systemctl restart tor@lnd

admin@bob:~ $ proxychains4 nmap -p9735 $(lncli getinfo|jq -c .uris|cut -d@ -f2|cut -d: -f1)
[proxychains] config file found: /etc/proxychains4.conf
[proxychains] preloading /usr/lib/aarch64-linux-gnu/libproxychains.so.4
[proxychains] DLL init: proxychains-ng 4.13
Starting Nmap 7.70 ( https://nmap.org ) at 2021-07-13 14:11 BST
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion:80 <--denied
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion:9735 <--denied
Nmap scan report for a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion (224.0.0.1)
Host is up (3.8s latency).
rDNS record for 224.0.0.1: all-systems.mcast.net

PORT     STATE  SERVICE
9735/tcp closed unknown

Nmap done: 1 IP address (1 host up) scanned in 5.49 seconds
```

@nyxnor your node should start with: 
```
PORT     STATE SERVICE
9735/tcp open  unknown
```
try restarting lnd last or a full reboot.
When the port is closed the node is not connectable leading to offline channels, force closes and general unreliability.

Next we can test if using `sudo pkill -sighup tor` only creates the new HiddenServices from the torrc and if yes it shouldreplace the `sudo systemctl restart tor@default` in `internet.hiddenservice.sh` (and other places) and we can try get rid of the `tor@lnd` instance.

**Comment by nyxnor** [comment]
>your node should start with:

Will need a full reboot, already restarted lnd and tor@lnd a few times now.

>Next we can test if using sudo pkill -sighup tor only creates the new HiddenServices from the torrc and if yes it shouldreplace the sudo systemctl restart tor@default in internet.hiddenservice.sh (and other places) and we can try get rid of the tor@lnd instance.

Yes it can and I am using that on my day to day use. Any config lines for the torrc that is added or replaced will be used after reloading tor, adding or removing hidden services/client authorization.

**Comment by openoms** [comment]
ok, the Hidden Service is created and functional without killing the lnd port so this should work.

```
admin@bob:~ $ sudo nano /etc/tor/instances/lnd/torrc 
admin@bob:~ $ sudo cat /mnt/hdd/tor-lnd/test/
cat: /mnt/hdd/tor-lnd/test/: No such file or directory
admin@bob:~ $ sudo cat /mnt/hdd/tor-lnd/test/hostname
cat: /mnt/hdd/tor-lnd/test/hostname: No such file or directory
admin@bob:~ $ sudo pkill -sighup tor
admin@bob:~ $ sudo cat /mnt/hdd/tor-lnd/test/hostname
ezn72yyes5dqgenmuqrv4zaov5kkj2zmwjukoaemeptnqnqqlytcrwad.onion
admin@bob:~ $ proxychains4 nmap -p9735 $(lncli getinfo|jq -c .uris|cut -d@ -f2|cut -d: -f1)
[proxychains] config file found: /etc/proxychains4.conf
[proxychains] preloading /usr/lib/aarch64-linux-gnu/libproxychains.so.4
[proxychains] DLL init: proxychains-ng 4.13
Starting Nmap 7.70 ( https://nmap.org ) at 2021-07-13 14:25 BST
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion:80 <--denied
[proxychains] Strict chain  ...  127.0.0.1:9050  ...  a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion:9735  ...  OK
Nmap scan report for a4w37h23fiium7anpy27ziflcnwb22l6okwtenjde2xmmf4icsawnuqd.onion (224.0.0.1)
Host is up (2.6s latency).
rDNS record for 224.0.0.1: all-systems.mcast.net

PORT     STATE SERVICE
9735/tcp open  unknown
Nmap done: 1 IP address (1 host up) scanned in 3.46 seconds
```

Next to weed out the `tor@lnd` instance and change over to `sudo pkill -sighup tor` everywhere.
Are you ok to have a look at that?
Will be happy to pitch in or review , but can be part of your PR.

**Comment by nyxnor** [comment]
>Are you ok to have a look at that?

Already did :), already coded, almost ready to review.

Keep this open until that PR is tested.

**Comment by xanoni** [comment]
By the way, I think the "official" way to do this is `systemctl reload tor` (instead of restart)

**Comment by nyxnor** [comment]
@xanoni I tested different methods to reload and documented the differences here https://github.com/rootzoll/raspiblitz/issues/2409#issuecomment-876340098
the reason I use `pkill -sighup` it because it is smaller (faster to type) than `systemctl reload`.

Edit: this works great when running headless, but beware that is recommended to use systemd for your desktop, as tor browser will stop working (you will have to close and open it again) when using pkill, as it gets the tor pid.

**Comment by xanoni** [comment]
OK. I just checked and the effect should be exactly the same though ... see `/etc/init.d/tor`:

```bash
211     if start-stop-daemon --stop --signal 1 --quiet --pidfile $TORPID --exec $DAEMON
212     then
213         log_action_end_msg 0
214     elif kill -0 $pid 2>/dev/null; then
215         log_action_end_msg 1 "Is $pid not $NAME?  Is $DAEMON a different binary now?"
216         exit 1
217     else
218         log_action_end_msg 1 "$DAEMON died: process $pid not running; or permission denied"
219         exit 1
```

I understand line 211 just sends a `HUP` (signal = 1).

**Comment by openoms** [comment]
ok @xanoni thanks I have tested this with c-lightning and `sudo systemctl reload tor` does not close the port. The question is if it creates the new hidden service?

See the sequence:
```
admin@raspberrypi:~ $ proxychains4 nmap -p19736 anibj3bogldf26ftlj2ap7brjo6drzoecgrc57fw3inmxdy3h7pa4nad.onion

PORT      STATE SERVICE
19736/tcp open  unknown

admin@raspberrypi:~ $ sudo systemctl reload tor
admin@raspberrypi:~ $ proxychains4 nmap -p19736 anibj3bogldf26ftlj2ap7brjo6drzoecgrc57fw3inmxdy3h7pa4nad.onion

PORT      STATE SERVICE
19736/tcp open  unknown

admin@raspberrypi:~ $ sudo systemctl restart tor
admin@raspberrypi:~ $ proxychains4 nmap -p19736 anibj3bogldf26ftlj2ap7brjo6drzoecgrc57fw3inmxdy3h7pa4nad.onion

PORT      STATE  SERVICE
19736/tcp closed unknown

admin@raspberrypi:~ $ sudo systemctl restart tlightningd
admin@raspberrypi:~ $ proxychains4 nmap -p19736 anibj3bogldf26ftlj2ap7brjo6drzoecgrc57fw3inmxdy3h7pa4nad.onion

PORT      STATE SERVICE
19736/tcp open  unknown
```

**Comment by openoms** [comment]
Great, the .onion is created too. See the test adding to torrc:
```
admin@raspberrypi:~ $ sudo nano /etc/tor/torrc 
admin@raspberrypi:~ $ sudo cat /mnt/hdd/tor/test/hostname
cat: /mnt/hdd/tor/test/hostname: No such file or directory
admin@raspberrypi:~ $ sudo systemctl reload tor
admin@raspberrypi:~ $ sudo cat /mnt/hdd/tor/test/hostname
fv56gfrfrus6tuzl6upo6degrexs6ip5zm2eqhon7ctgojcihdo2taad.onion
```

@nyxnor seems that  
`sudo systemctl reload tor`
has the same effect as
`sudo pkill -sighup tor`.

As they seem to be equivalent in function I see no reason to change this in #2419 but good to be aware.
Edited my comment in: https://github.com/lightningnetwork/lnd/issues/5166#issuecomment-880152737 too.

**Comment by xanoni** [comment]
Yes it really shouldn't make a difference. It's just a syntax difference, unless Tor changes how it triggers config reloads (unlikely, given HUP is very standard). Using `systemctl` is a bit more universal, though, because should something change, then the Debian team would update the `reload` command for us :-)

**Comment by nyxnor** [comment]
@openoms 
> As they seem to be equivalent in function I see no reason to change this in #2419 but good to be aware.

It is here commented https://github.com/nyxnor/raspiblitz/blob/tor-patch/home.admin/config.scripts/tor.functions.lib#L186, will update.

**Comment by rootzoll** [comment]
@openoms can you summarize the state of this issue? is there something for the v1.7.1 needed?

**Comment by nyxnor** [comment]
@rootzoll he already did https://github.com/rootzoll/raspiblitz/commit/4874d7dd15c936e3bbd3c5cc8a0ceeba1c5abf69#diff-eb88b0a6c4c9512eeda4cdba40f682e1cb07504afd895dc465191da40ddc09a4

@openoms I am using `sudo systemctl reload-or-restart tor@default` because if reload fails because of some old misconfiguration in the torrc, it will restart. Also likewhoa reported having issues only reloading tor https://code.samourai.io/ronindojo/RoninDojo/-/issues/97, but that was on their distro and I could not reproduce.

**Comment by openoms** [comment]
@rootzoll yes basically the separate `lnd@tor` instance now removed because changing 
`sudo systemctl restart tor@default`
to
`sudo systemctl reload tor@default`
solves the issue as tested.

So changes are:
* reloading Tor instead of restarting on `torrc` changes
* using only one tor instance now: `tor@default`

The lnd tor streamisolation remains active which opens new circuits for all the lnd connections 
There is no similar (streamisolation) option for C-lightning so introducing a dedicated Tor instance could be considered.

**Comment by openoms** [comment]
Since we are not restarting Tor any more this issue can be closed.

**Comment by xanoni** [comment]
Sorry to be contrarian here, but I actually set up a separate Tor instance for my `lnd` after seeing it in this repo and I'm still using it :) It seems it helped me fix some channel stability issues.

However, I think this is because I have a "shit ton" of other things using Tor on the same Pi. Probably not an issue for normal RaspiBlitz users. Just FYI.

**Comment by openoms** [comment]
@xanoni see this https://github.com/rootzoll/raspiblitz/issues/2560 for further discussion about the possible benefits of reintroducing separate Tor instances for the Lightning nodes.
